### PR TITLE
add better support for IWYU and fix things

### LINKED ISSUE #2957: fold include-what-you-use into `make lint`
I really like ["Include What You Use"](https://github.com/include-what-you-use/include-what-you-use). @zanchey added support to the makefile in commit 299a383d on 2015-07-29. The problem with that change from my perspective is that the `make iwyu` target doesn't limit itself to the files being modified; or were modified by the most recent commit. I propose folding that functionality into the `lint` and `lint-all` targets I recently introduced.

Has anyone been running `make iwyu`? I ask because when I run that on my system it produces a large number of recommendations to improve the code.

### PR DESCRIPTION
**Pull request by krader1961**
_This change was far more difficult than I anticipated. It also does not remove all IWYU errors as I had hoped to accomplish when I embarked on this initiative. What I want from people at this point is to verify this change does not break `make test` and that `make install` yields a usable fish shell. Both of those are true for me on OS X and Ubuntu 14.04. Once this change is merged we can collectively work to remove the remaining IWYU errors. Also, I explicitly did not restyle the code I touched to match the new style guide since it was obvious the change was already so big that introducing style changes was a bad idea.

What follows is the commit comment._

Remove the "make iwyu" build target. Move the functionality into the
recently introduced lint.fish script. Fix a lot, but not all, of the
include-what-you-use errors. Specifically, it fixes all of the IWYU errors
on my OS X server but only removes some of them on my Ubuntu 14.04 server.

Fixes #2957

### CODE REVIEW

**Comment by ridiculousfish** [comment]
Reviewed with a lot of help from grep, changes look good to me. make test works, make install yields a usable shell. Nice work!

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,9 +1,20 @@
+### Import What You Use
```
Comment (floam) [inline CONTRIBUTING.md]: Likely brain-o: `s/import/include`

**Comment by zanchey** [comment]
Works on FreeBSD 10.2 as well.

**Comment by krader1961** [comment]
Reverted the change to .gitignore and completed documenting IWYU in CONTRIBUTING.md. Merged as 1f06e5f0b9ee483053b987c9cab9f1f5fce2590c.

**Pull request closed without merging**